services/user/Chainmail/plugin/src/lib.rs

```diff
@@ -22,6 +32,35 @@ impl API for ChainmailPlugin {
         )?;
         Ok(())
     }
+
+    fn archive(event_id: u64) -> Result<(), Error> {
+        Transact::add_action_to_transaction(
+            "archive",
+            &chainmail::action_structs::archive { event_id }.packed(),
+        )?;
+        Ok(())
+    }
+
+    fn save(event_id: u64) -> Result<(), Error> {
+        // look up message details via event_id
+        // let (sender, receiver, subject, body) = fetch.get(/rest/message by id);
+        let res = server::get_json(&format!("/messages?id={}", event_id))?;
+
+        let msg = serde_json::from_str::<Message>(&res)
+            .map_err(|err| ErrorType::QueryResponseParseError.err(err.to_string().as_str()))?;
+
+        // save the message to state
+        Transact::add_action_to_transaction(
+            "save",
+            &chainmail::action_structs::save {
+                receiver: msg.receiver,
```

Comment:
If you're saving a sent message, then it makes sense to save the receiver. If you're saving a received message, then it makes sense to save the sender. So you probably want to always save both sender + receiver.